server/src/main/java/com/tencent/rss/server/buffer/ShuffleBuffer.java

```diff
@@ -0,0 +1,285 @@
+  public synchronized ShuffleDataResult getShuffleData(
+      long lastBlockId, int readBufferSize) {
+    try {
+      List<BufferSegment> bufferSegments = Lists.newArrayList();
+      List<ShufflePartitionedBlock> readBlocks = Lists.newArrayList();
+      updateBufferSegmentsAndResultBlocks(
+          lastBlockId, readBufferSize, bufferSegments, readBlocks);
+      if (!bufferSegments.isEmpty()) {
+        BufferSegment bufferSegment = bufferSegments.get(bufferSegments.size() - 1);
+        int length = bufferSegment.getOffset() + bufferSegment.getLength();
+        byte[] data = new byte[length];
+        // copy result data
+        updateShuffleData(readBlocks, data);
+        return new ShuffleDataResult(data, bufferSegments);
+      }
+    } catch (Exception e) {
+      LOG.error("Exception happened when getShuffleData in buffer", e);
+    }
+    return new ShuffleDataResult();
+  }
+
+  // here is the rule to read data in memory:
+  // 1. read from inFlushBlockMap order by eventId asc, then from blocks
+  // 2. if can't find lastBlockId, means related data may be flushed to storage, repeat step 1
+  private void updateBufferSegmentsAndResultBlocks(
+      long lastBlockId,
+      long readBufferSize,
+      List<BufferSegment> bufferSegments,
+      List<ShufflePartitionedBlock> resultBlocks) {
+    long nextBlockId = lastBlockId;
+    List<Long> sortedEventId = sortFlushingEventId();
+    int offset = 0;
+    boolean hasLastBlockId = false;
+    // read from inFlushBlockMap first to make sure the order of
+    // data read is according to the order of data received
+    if (!inFlushBlockMap.isEmpty()) {
+      for (Long eventId : sortedEventId) {
+        // update bufferSegments with current blocks in flush map
+        CachedBlocksReadInfo cachedBlocksReadInfo = updateBufferSegmentsWithCachedBlocks(
+            offset, inFlushBlockMap.get(eventId), nextBlockId, readBufferSize,
+            bufferSegments, resultBlocks);
+        offset = cachedBlocksReadInfo.getOffsetInResultData();
+        // if has data or find last blockId, read from begin with next cached blocks
+        if (offset > 0 || cachedBlocksReadInfo.isHasLastBlockId()) {
+          // read from begin in next cached blocks
+          nextBlockId = Constants.INVALID_BLOCK_ID;
+          hasLastBlockId = true;
+        }
+        if (offset >= readBufferSize) {
+          break;
+        }
+      }
+    }
+    // try to read from cached blocks which is not in flush queue
+    if (blocks.size() > 0 && offset < readBufferSize) {
+      CachedBlocksReadInfo cachedBlocksReadInfo = updateBufferSegmentsWithCachedBlocks(
+          offset, blocks, nextBlockId, readBufferSize, bufferSegments, resultBlocks);
+      offset = cachedBlocksReadInfo.getOffsetInResultData();
+      hasLastBlockId = cachedBlocksReadInfo.isHasLastBlockId();
+    }
+    if (lastBlockId != Constants.INVALID_BLOCK_ID && offset == 0 && !hasLastBlockId) {
```

Comment:
`offset == 0` is enough?

Reply:
offset = 0 means there has no data, if lastBlockId != -1,  here is the different case:
1. can't find lastBlockId, we should read from begin
2. find lastBlockId, it is the last block, just return empty

Reply:
OK